Tagged Annotations

BryonLewis opened the pull request:
This is a PR to introduce Tagged Annotations as a precursor to #1377.

- Adds a capability to save and load annotations with a 'tag' which is a user-defined string for a set of annotations.

TODO:

- [x] Add UI to see a list of available tags for dataset.  By default it will use the untagged version
- [x] The UI should have an option to load up a tagged version.  Probably somewhere close to the title of the dataset.
- [x] The Revision history should filter out the tags that don't match the currently tagged dataset.
- [x] When saving there should be an option to use the existing tag or create a new tag from the current annotations
- [x] Add URL parameter to router to automatically set the loaded annotation.  This may be an array because in the future we may want to compare two sets of annotations.

Possible Questions/Complications:

- Running Training/Pipelines will be done on the main annotations and not a tagged set for now.
- Not working on the Desktop Version of tags right now.  I think it would include a concept of different JSON files that could be used.

https://github.com/Kitware/dive/assets/61746913/baa62d88-483c-4ce6-8116-7068f33509c9

BryonLewis commented:
TODO:

- [ ] Tooltips for the current tag, and the sidepanel chevron
- [ ] Move the selection of the current tag into a sidepanel context
- [ ] The sid panel context will also have the option to select new tags or to create a new tag by saving as.... (This would mean removing the chevron from the current Save).
- [ ] Import should only show the tag option if there are multiple tags already.

BryonLewis commented:
Updated Demo
https://github.com/Kitware/dive/assets/61746913/74829bf4-fb7b-4079-ba1f-0c8bde3668aa

marySalvi commented on client/platform/web-girder/views/AnnotationTags.vue:
```diff
@@ -0,0 +1,89 @@
+
+<template>
+  <div>
+    <p> Available Tags</p>
```
```suggestion
        <div class="px-4 pt-2 text-body-1"> Available Tags</div>
```

This would align the text and improve the spacing.

marySalvi commented on client/platform/web-girder/views/AnnotationTags.vue:
```diff
@@ -0,0 +1,89 @@
+              :disabled="!validForm"
+              @click="addTag()"
+            >
+              Add Tag
```
Currently you can add a tag even when there are no annotations. Is that expected behavior?

BryonLewis replied:
Yeah, a Tag doesn't require there to be annotations in it.  It's just when you generate a new tag I figured a reasonable default behavior would be to copy over the existing annotations.

marySalvi commented on client/platform/web-girder/views/AnnotationTags.vue:
```diff
@@ -0,0 +1,89 @@
+      <v-expansion-panel>
+        <v-expansion-panel-header> Add New Tag</v-expansion-panel-header>
+        <v-expansion-panel-content>
+          <p>
+            Add a new Tag with a custom name that can be used to reference it in the future.
+            'default' is a reserved tag which can't be used.
+            Adding a new tag will tag the current annotations and copy them to the new tag.
+          </p>
```
```suggestion
          <p>
          Enter a name for a new annotation group. Grouping annotations allows for easy comparison. Creating a new 
          group will add all current annotations to that group.
          </p>
```

I do think we need to change the name from tag to group or collection or something like that. Currently new annotations get added to the selected tag as well. Should this be included here or is that intuitive enough?

BryonLewis replied:
Can't use group, we already have that reserved for groupings of tracks.  Collection has a meaning in a girder context.

marySalvi replied:
batch?

BryonLewis replied:
If we're going towards slicer CLI in the future, the girder plugin has terminology using 'batch' for running the same job on multiple datasets/items.  I don't know if that concept overlaps completely with what is done here to use a similar name or if they are two different and should use different names to prevent confusion.
https://github.com/girder/slicer_cli_web/blob/2cfbb92256f840e1edd3004632432ce1d0634209/README.rst#batch-processing

marySalvi replied:
Discussion offline concluded in using 'Annotation-Set'

marySalvi commented on client/platform/web-girder/views/AnnotationTags.vue:
```diff
@@ -0,0 +1,89 @@
+      >
+        <v-chip
+          outlined
+          :color="tag === currentTag || (!currentTag && tag === 'default') ? 'cyan' : 'white'"
```
I think when someone navigates here it should show the `default` tag. It was a little surprising to have two tags appear when saving the first. Also then we likely wouldn't need to alert the user not to use default because its clearly already used.

BryonLewis replied:
I'll add that in so that on initial navigation even with no tags it shows the 'default' tag in the list.

marySalvi commented on client/platform/web-girder/views/AnnotationTags.vue:
```diff
@@ -0,0 +1,89 @@
+          <p>
+            Add a new Tag with a custom name that can be used to reference it in the future.
+            'default' is a reserved tag which can't be used.
+            Adding a new tag will tag the current annotations and copy them to the new tag.
```
I wonder if it should be an option to add current annotations to a tag. I think I would expect the initial annotations to belong to `default`. I am thinking of `default` as ground truth, is that an incorrect thought?

I would expect everything after to be part of the new tag but not _necessarily_ everything before.

BryonLewis replied:
So if you have no annotations at all there is no tag.  Once you start adding tags the previous annotations had an empty tag I named 'default'.  'default' can be whatever you want annotation for you annotations but it's the tag that pipelines/training will run on.  

If you update the annotations in tag and click save it will save to the currently selected tag.  I didn't implement anything for transferring annotations between tags except in the case of generating a brand new tag.

BryonLewis commented:
TODO for addressing comments:

- [ ] Rename 'tags' to 'Annotation Set'
- [ ] Add default to the base view when launching the Annotation Set window.

Subsequently need to update the track comparison branch afterwards.

marySalvi reviewed (approved):
LGTM!
